File: leetcode/python/337. 打家劫舍 III.py
#!/bin/bash python
"""
337. 打家劫舍 III

思路：

树形dp
每棵子树返回两个状态值，当前节点被偷/不偷后向下得到的最大值
当前根节点更加左右子树的不同状态，返回root被偷/不偷后的最大值
"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def rob(self, root: TreeNode) -> int:
        """
        树形dp
        截止到当前root节点，能偷到的最大钱与两个子树的钱有关
        每个子树都应该返回，当前被偷或者不偷，能得到的最大钱
        """
        def back(root):
            """
            后续遍历
            """
            if root == None:
                return 0, 0
            leftOn, leftOff = back(root.left)
            rightOn, rightOff = back(root.right)
            rootOn = root.val + leftOff + rightOff
            rootOff = max(leftOn, leftOff) + max(rightOn, rightOff)
            return rootOn, rootOff
        return max(back(root))

File: leetcode/cpp/337. 打家劫舍 III.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
#define MAX(x, y) ((x) > (y) ? (x) : (y))
class Solution {
public:
    // 后续遍历
    void back(TreeNode* root, int& rootOn, int& rootOff) {
        if (root == nullptr) {
            return;
        }
        int leftOn = 0, leftOff = 0, rightOn = 0, rightOff = 0;
        back(root->left, leftOn, leftOff);
        back(root->right, rightOn, rightOff);
        rootOn = root->val + leftOff + rightOff;
        rootOff = MAX(leftOn, leftOff) + MAX(rightOn, rightOff);
        return;
    }
	
    int rob(TreeNode* root) {
        int rootOn = 0, rootOff = 0;
        back(root, rootOn, rootOff);
        return MAX(rootOn, rootOff);
    }
};